File: python/semantic_kernel/agents/orchestration/group_chat.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio
import inspect
import logging
import sys
from abc import ABC, abstractmethod
from collections.abc import Awaitable, Callable
from typing import Generic, TypeVar

from semantic_kernel.agents.agent import Agent
from semantic_kernel.agents.orchestration.agent_actor_base import ActorBase, AgentActorBase
from semantic_kernel.agents.orchestration.orchestration_base import DefaultTypeAlias, OrchestrationBase, TIn, TOut
from semantic_kernel.agents.runtime.core.cancellation_token import CancellationToken
from semantic_kernel.agents.runtime.core.core_runtime import CoreRuntime
from semantic_kernel.agents.runtime.core.message_context import MessageContext
from semantic_kernel.agents.runtime.core.routed_agent import message_handler
from semantic_kernel.agents.runtime.core.topic import TopicId
from semantic_kernel.agents.runtime.in_process.type_subscription import TypeSubscription
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.contents.chat_message_content import ChatMessageContent
from semantic_kernel.contents.streaming_chat_message_content import StreamingChatMessageContent
from semantic_kernel.contents.utils.author_role import AuthorRole
from semantic_kernel.kernel_pydantic import KernelBaseModel
from semantic_kernel.utils.feature_stage_decorator import experimental

if sys.version_info >= (3, 12):
    from typing import override  # pragma: no cover
else:
    from typing_extensions import override  # pragma: no cover


logger: logging.Logger = logging.getLogger(__name__)


# region Messages and Types


@experimental
class GroupChatStartMessage(KernelBaseModel):
    """A message type to start a group chat."""

    body: DefaultTypeAlias


@experimental
class GroupChatRequestMessage(KernelBaseModel):
    """A request message type for agents in a group chat."""

    agent_name: str


@experimental
class GroupChatResponseMessage(KernelBaseModel):
    """A response message type from agents in a group chat."""

    body: ChatMessageContent


_TGroupChatManagerResult = TypeVar("_TGroupChatManagerResult", ChatMessageContent, str, bool)


@experimental
class GroupChatManagerResult(KernelBaseModel, Generic[_TGroupChatManagerResult]):
    """A result message type from the group chat manager."""

    result: _TGroupChatManagerResult
    reason: str


# Subclassing GroupChatManagerResult to create specific result types because
# we need to change the names of the classes to remove the generic type parameters.
# Many model services (e.g. OpenAI) do not support generic type parameters in the
# class name (e.g. "GroupChatManagerResult[bool]").
@experimental
class BooleanResult(GroupChatManagerResult[bool]):
    """A result message type from the group chat manager with a boolean result."""

    pass


@experimental
class StringResult(GroupChatManagerResult[str]):
    """A result message type from the group chat manager with a string result."""

    pass


@experimental
class MessageResult(GroupChatManagerResult[ChatMessageContent]):
    """A result message type from the group chat manager with a message result."""

    pass


# endregion Messages and Types

# region GroupChatAgentActor


@experimental
class GroupChatAgentActor(AgentActorBase):
    """An agent actor that process messages in a group chat."""

    @message_handler
    async def _handle_start_message(self, message: GroupChatStartMessage, ctx: MessageContext) -> None:
        """Handle the start message for the group chat."""
        logger.debug(f"{self.id}: Received group chat start message.")
        if isinstance(message.body, ChatMessageContent):
            if self._agent_thread:
                await self._agent_thread.on_new_message(message.body)
            else:
                self._chat_history.add_message(message.body)
        elif isinstance(message.body, list) and all(isinstance(m, ChatMessageContent) for m in message.body):
            if self._agent_thread:
                for m in message.body:
                    await self._agent_thread.on_new_message(m)
            else:
                for m in message.body:
                    self._chat_history.add_message(m)
        else:
            raise ValueError(f"Invalid message body type: {type(message.body)}. Expected {DefaultTypeAlias}.")

    @message_handler
    async def _handle_response_message(self, message: GroupChatResponseMessage, ctx: MessageContext) -> None:
        logger.debug(f"{self.id}: Received group chat response message.")
        if self._agent_thread is not None:
            await self._agent_thread.on_new_message(message.body)
        else:
            self._chat_history.add_message(message.body)

    @message_handler
    async def _handle_request_message(self, message: GroupChatRequestMessage, ctx: MessageContext) -> None:
        if message.agent_name != self._agent.name:
            return

        logger.debug(f"{self.id}: Received group chat request message.")

        response = await self._invoke_agent()

        logger.debug(f"{self.id} responded with {response}.")

        await self.publish_message(
            GroupChatResponseMessage(body=response),
            TopicId(self._internal_topic_type, self.id.key),
            cancellation_token=ctx.cancellation_token,
        )


# endregion GroupChatAgentActor


# region GroupChatManager


@experimental
class GroupChatManager(KernelBaseModel, ABC):
    """A group chat manager that manages the flow of a group chat."""

    current_round: int = 0
    max_rounds: int | None = None

    human_response_function: Callable[[ChatHistory], Awaitable[ChatMessageContent] | ChatMessageContent] | None = None

    @abstractmethod
    async def should_request_user_input(self, chat_history: ChatHistory) -> BooleanResult:
        """Check if the group chat should request user input.

        Args:
            chat_history (ChatHistory): The chat history of the group chat.
        """
        ...

    async def should_terminate(self, chat_history: ChatHistory) -> BooleanResult:
        """Check if the group chat should terminate.

        Args:
            chat_history (ChatHistory): The chat history of the group chat.
        """
        self.current_round += 1

        if self.max_rounds is not None:
            return BooleanResult(
                result=self.current_round > self.max_rounds,
                reason="Maximum rounds reached."
                if self.current_round > self.max_rounds
                else "Not reached maximum rounds.",
            )
        return BooleanResult(result=False, reason="No maximum rounds set.")

    @abstractmethod
    async def select_next_agent(
        self,
        chat_history: ChatHistory,
        participant_descriptions: dict[str, str],
    ) -> StringResult:
        """Select the next agent to speak.

        Args:
            chat_history (ChatHistory): The chat history of the group chat.
            participant_descriptions (dict[str, str]): The descriptions of the participants in the group chat.
        """
        ...

    @abstractmethod
    async def filter_results(
        self,
        chat_history: ChatHistory,
    ) -> MessageResult:
        """Filter the results of the group chat.

        Args:
            chat_history (ChatHistory): The chat history of the group chat.
            participant_descriptions (dict[str, str]): The descriptions of the participants in the group chat.
        """
        ...


@experimental
class RoundRobinGroupChatManager(GroupChatManager):
    """A round-robin group chat manager."""

    current_index: int = 0

    @override
    async def should_request_user_input(self, chat_history: ChatHistory) -> BooleanResult:
        """Check if the group chat should request user input."""
        return BooleanResult(
            result=False,
            reason="The default round-robin group chat manager does not request user input.",
        )

    @override
    async def select_next_agent(
        self,
        chat_history: ChatHistory,
        participant_descriptions: dict[str, str],
    ) -> StringResult:
        """Select the next agent to speak."""
        next_agent = list(participant_descriptions.keys())[self.current_index]
        self.current_index = (self.current_index + 1) % len(participant_descriptions)
        return StringResult(result=next_agent, reason="Round-robin selection.")

    @override
    async def filter_results(
        self,
        chat_history: ChatHistory,
    ) -> MessageResult:
        """Filter the results of the group chat."""
        return MessageResult(
            result=chat_history.messages[-1],
            reason="The last message in the chat history is the result in the default round-robin group chat manager.",
        )


# endregion GroupChatManager

# region GroupChatManagerActor


@experimental
class GroupChatManagerActor(ActorBase):
    """A group chat manager actor."""

    def __init__(
        self,
        manager: GroupChatManager,
        internal_topic_type: str,
        participant_descriptions: dict[str, str],
        result_callback: Callable[[DefaultTypeAlias], Awaitable[None]] | None = None,
    ):
        """Initialize the group chat manager actor.

        Args:
            manager (GroupChatManager): The group chat manager that manages the flow of the group chat.
            internal_topic_type (str): The topic type of the internal topic.
            participant_descriptions (dict[str, str]): The descriptions of the participants in the group chat.
            agent_response_callback (Callable | None): A function that is called when a response is produced
                by the agents.
            result_callback (Callable | None): A function that is called when the group chat manager produces a result.
        """
        self._manager = manager
        self._internal_topic_type = internal_topic_type
        self._chat_history = ChatHistory()
        self._participant_descriptions = participant_descriptions
        self._result_callback = result_callback

        super().__init__(description="An actor for the group chat manager.")

    @message_handler
    async def _handle_start_message(self, message: GroupChatStartMessage, ctx: MessageContext) -> None:
        """Handle the start message for the group chat."""
        logger.debug(f"{self.id}: Received group chat start message.")
        if isinstance(message.body, ChatMessageContent):
            self._chat_history.add_message(message.body)
        elif isinstance(message.body, list) and all(isinstance(m, ChatMessageContent) for m in message.body):
            for m in message.body:
                self._chat_history.add_message(m)
        else:
            raise ValueError(f"Invalid message body type: {type(message.body)}. Expected {DefaultTypeAlias}.")

        await self._determine_state_and_take_action(ctx.cancellation_token)

    @message_handler
    async def _handle_response_message(self, message: GroupChatResponseMessage, ctx: MessageContext) -> None:
        if message.body.role != AuthorRole.USER:
            self._chat_history.add_message(
                ChatMessageContent(
                    role=AuthorRole.USER,
                    content=f"Transferred to {message.body.name}",
                )
            )
        self._chat_history.add_message(message.body)

        await self._determine_state_and_take_action(ctx.cancellation_token)

    async def _determine_state_and_take_action(self, cancellation_token: CancellationToken) -> None:
        """Determine the state of the group chat and take action accordingly."""
        # User input state
        should_request_user_input = await self._manager.should_request_user_input(
            self._chat_history.model_copy(deep=True)
        )
        if should_request_user_input.result and self._manager.human_response_function:
            logger.debug(f"Group chat manager requested user input. Reason: {should_request_user_input.reason}")
            user_input_message = await self._call_human_response_function()
            self._chat_history.add_message(user_input_message)
            await self.publish_message(
                GroupChatResponseMessage(body=user_input_message),
                TopicId(self._internal_topic_type, self.id.key),
                cancellation_token=cancellation_token,
            )
            logger.debug("User input received and added to chat history.")

        # Determine if the group chat should terminate
        should_terminate = await self._manager.should_terminate(self._chat_history.model_copy(deep=True))
        if should_terminate.result:
            logger.debug(f"Group chat manager decided to terminate the group chat. Reason: {should_terminate.reason}")
            if self._result_callback:
                result = await self._manager.filter_results(self._chat_history.model_copy(deep=True))
                result.result.metadata["termination_reason"] = should_terminate.reason
                result.result.metadata["filter_result_reason"] = result.reason
                await self._result_callback(result.result)
            return

        # Select the next agent to speak if the group chat is not terminating
        next_agent = await self._manager.select_next_agent(
            self._chat_history.model_copy(deep=True),
            self._participant_descriptions,
        )
        logger.debug(
            f"Group chat manager selected agent: {next_agent.result} on round {self._manager.current_round}. "
            f"Reason: {next_agent.reason}"
        )

        await self.publish_message(
            GroupChatRequestMessage(agent_name=next_agent.result),
            TopicId(self._internal_topic_type, self.id.key),
            cancellation_token=cancellation_token,
        )

    async def _call_human_response_function(self) -> ChatMessageContent:
        """Call the human response function if it is set."""
        assert self._manager.human_response_function  # nosec B101
        if inspect.iscoroutinefunction(self._manager.human_response_function):
            return await self._manager.human_response_function(self._chat_history.model_copy(deep=True))
        return self._manager.human_response_function(self._chat_history.model_copy(deep=True))  # type: ignore[return-value]


# endregion GroupChatManagerActor

# region GroupChatOrchestration


@experimental
class GroupChatOrchestration(OrchestrationBase[TIn, TOut]):
    """A group chat multi-agent pattern orchestration."""

    def __init__(
        self,
        members: list[Agent],
        manager: GroupChatManager,
        name: str | None = None,
        description: str | None = None,
        input_transform: Callable[[TIn], Awaitable[DefaultTypeAlias] | DefaultTypeAlias] | None = None,
        output_transform: Callable[[DefaultTypeAlias], Awaitable[TOut] | TOut] | None = None,
        agent_response_callback: Callable[[DefaultTypeAlias], Awaitable[None] | None] | None = None,
        streaming_agent_response_callback: Callable[[StreamingChatMessageContent, bool], Awaitable[None] | None]
        | None = None,
    ) -> None:
        """Initialize the group chat orchestration.

        Args:
            members (list[Agent | OrchestrationBase]): A list of agents or orchestrations that are part of the
                handoff group. This first agent in the list will be the one that receives the first message.
            manager (GroupChatManager): The group chat manager that manages the flow of the group chat.
            name (str | None): The name of the orchestration.
            description (str | None): The description of the orchestration.
            input_transform (Callable | None): A function that transforms the external input message.
            output_transform (Callable | None): A function that transforms the internal output message.
            agent_response_callback (Callable | None): A function that is called when a full response is produced
                by the agents.
            streaming_agent_response_callback (Callable | None): A function that is called when a streaming response
                is produced by the agents.
        """
        self._manager = manager

        for member in members:
            if member.description is None:
                raise ValueError("All members must have a description.")

        super().__init__(
            members=members,
            name=name,
            description=description,
            input_transform=input_transform,
            output_transform=output_transform,
            agent_response_callback=agent_response_callback,
            streaming_agent_response_callback=streaming_agent_response_callback,
        )

    @override
    async def _start(
        self,
        task: DefaultTypeAlias,
        runtime: CoreRuntime,
        internal_topic_type: str,
        cancellation_token: CancellationToken,
    ) -> None:
        """Start the group chat process.

        This ensures that all initial messages are sent to the individual actors
        and processed before the group chat begins. It's important because if the
        manager actor processes its start message too quickly (or other actors are
        too slow), it might send a request to the next agent before the other actors
        have the necessary context.
        """

        async def send_start_message(agent: Agent) -> None:
            target_actor_id = await runtime.get(self._get_agent_actor_type(agent, internal_topic_type))
            await runtime.send_message(
                GroupChatStartMessage(body=task),
                target_actor_id,
                cancellation_token=cancellation_token,
            )

        await asyncio.gather(*[send_start_message(agent) for agent in self._members])

        # Send the start message to the manager actor
        target_actor_id = await runtime.get(self._get_manager_actor_type(internal_topic_type))
        await runtime.send_message(
            GroupChatStartMessage(body=task),
            target_actor_id,
            cancellation_token=cancellation_token,
        )

    @override
    async def _prepare(
        self,
        runtime: CoreRuntime,
        internal_topic_type: str,
        result_callback: Callable[[DefaultTypeAlias], Awaitable[None]],
    ) -> None:
        """Register the actors and orchestrations with the runtime and add the required subscriptions."""
        await self._register_members(runtime, internal_topic_type)
        await self._register_manager(runtime, internal_topic_type, result_callback=result_callback)
        await self._add_subscriptions(runtime, internal_topic_type)

    async def _register_members(self, runtime: CoreRuntime, internal_topic_type: str) -> None:
        """Register the agents."""
        await asyncio.gather(*[
            GroupChatAgentActor.register(
                runtime,
                self._get_agent_actor_type(agent, internal_topic_type),
                lambda agent=agent: GroupChatAgentActor(  # type: ignore[misc]
                    agent,
                    internal_topic_type,
                    agent_response_callback=self._agent_response_callback,
                    streaming_agent_response_callback=self._streaming_agent_response_callback,
                ),
            )
            for agent in self._members
        ])

    async def _register_manager(
        self,
        runtime: CoreRuntime,
        internal_topic_type: str,
        result_callback: Callable[[DefaultTypeAlias], Awaitable[None]] | None = None,
    ) -> None:
        """Register the group chat manager."""
        await GroupChatManagerActor.register(
            runtime,
            self._get_manager_actor_type(internal_topic_type),
            lambda: GroupChatManagerActor(
                self._manager,
                internal_topic_type=internal_topic_type,
                participant_descriptions={agent.name: agent.description for agent in self._members},  # type: ignore[misc]
                result_callback=result_callback,
            ),
        )

    async def _add_subscriptions(self, runtime: CoreRuntime, internal_topic_type: str) -> None:
        """Add subscriptions."""
        subscriptions: list[TypeSubscription] = []
        for agent in self._members:
            subscriptions.append(
                TypeSubscription(internal_topic_type, self._get_agent_actor_type(agent, internal_topic_type))
            )
        subscriptions.append(TypeSubscription(internal_topic_type, self._get_manager_actor_type(internal_topic_type)))

        await asyncio.gather(*[runtime.add_subscription(sub) for sub in subscriptions])

    def _get_agent_actor_type(self, agent: Agent, internal_topic_type: str) -> str:
        """Get the actor type for an agent.

        The type is appended with the internal topic type to ensure uniqueness in the runtime
        that may be shared by multiple orchestrations.
        """
        return f"{agent.name}_{internal_topic_type}"

    def _get_manager_actor_type(self, internal_topic_type: str) -> str:
        """Get the actor type for the group chat manager.

        The type is appended with the internal topic type to ensure uniqueness in the runtime
        that may be shared by multiple orchestrations.
        """
        return f"{GroupChatManagerActor.__name__}_{internal_topic_type}"


# endregion GroupChatOrchestration

File: nodejs/semantic-kernel/agents/orchestration/group-chat.ts
import { ChatHistory } from '../../contents/chat-history'
import { ChatMessageContent } from '../../contents/chat-message-content'
import { StreamingChatMessageContent } from '../../contents/streaming-chat-message-content'
import { AuthorRole } from '../../contents/utils/author-role'
import { experimental } from '../../utils/feature-stage-decorator'
import { createDefaultLogger } from '../../utils/logger'
import { Agent } from '../agent'
import { CancellationToken } from '../runtime/core/cancellation-token'
import { CoreRuntime } from '../runtime/core/core-runtime'
import { MessageContext } from '../runtime/core/message-context'
import { event } from '../runtime/core/routed-agent'
import { TopicId } from '../runtime/core/topic'
import { TypeSubscription } from '../runtime/in-process/type-subscription'
import { ActorBase, AgentActorBase } from './agent-actor-base'
import type { DefaultTypeAlias } from './orchestration-base'
import { OrchestrationBase } from './orchestration-base'

const logger = createDefaultLogger('GroupChat')

// region Messages and Types

/**
 * A message type to start a group chat.
 */
@experimental
export class GroupChatStartMessage {
  body: DefaultTypeAlias

  constructor(body: DefaultTypeAlias) {
    this.body = body
  }
}

/**
 * A request message type for agents in a group chat.
 */
@experimental
export class GroupChatRequestMessage {
  agentName: string

  constructor(agentName: string) {
    this.agentName = agentName
  }
}

/**
 * A response message type from agents in a group chat.
 */
@experimental
export class GroupChatResponseMessage {
  body: ChatMessageContent

  constructor(body: ChatMessageContent) {
    this.body = body
  }
}

/**
 * A result message type from the group chat manager.
 */
@experimental
export class GroupChatManagerResult<T> {
  result: T
  reason: string

  constructor(result: T, reason: string) {
    this.result = result
    this.reason = reason
  }
}

/**
 * A result message type from the group chat manager with a boolean result.
 */
@experimental
export class BooleanResult extends GroupChatManagerResult<boolean> {
  constructor(result: boolean, reason: string) {
    super(result, reason)
  }
}

/**
 * A result message type from the group chat manager with a string result.
 */
@experimental
export class StringResult extends GroupChatManagerResult<string> {
  constructor(result: string, reason: string) {
    super(result, reason)
  }
}

/**
 * A result message type from the group chat manager with a message result.
 */
@experimental
export class MessageResult extends GroupChatManagerResult<ChatMessageContent> {
  constructor(result: ChatMessageContent, reason: string) {
    super(result, reason)
  }
}

// endregion Messages and Types

// region GroupChatAgentActor

/**
 * An agent actor that process messages in a group chat.
 */
@experimental
export class GroupChatAgentActor extends AgentActorBase {
  @event()
  async _handleStartMessage(message: GroupChatStartMessage, _ctx: MessageContext): Promise<void> {
    logger.debug(`${this.id}: Received group chat start message.`)
    if (message.body instanceof ChatMessageContent) {
      this._messageCache.addMessage(message.body)
    } else if (Array.isArray(message.body) && message.body.every((m) => m instanceof ChatMessageContent)) {
      for (const m of message.body) {
        this._messageCache.addMessage(m)
      }
    } else {
      throw new Error(`Invalid message body type: ${typeof message.body}. Expected DefaultTypeAlias.`)
    }
  }

  @event()
  async _handleResponseMessage(message: GroupChatResponseMessage, _ctx: MessageContext): Promise<void> {
    logger.debug(`${this.id}: Received group chat response message.`)
    this._messageCache.addMessage(message.body)
  }

  @event()
  async _handleRequestMessage(message: GroupChatRequestMessage, ctx: MessageContext): Promise<void> {
    if (message.agentName !== this._agent.name) {
      return
    }

    logger.debug(`${this.id}: Received group chat request message.`)

    const response = await this._invokeAgent()

    logger.debug(`${this.id} responded with ${response}.`)

    await this.publishMessage(
      new GroupChatResponseMessage(response),
      new TopicId(this._internalTopicType, this.id.key),
      { cancellationToken: ctx.cancellationToken }
    )
  }
}

// Add message type registrations for the handlers
;(GroupChatAgentActor.prototype._handleStartMessage as any).targetTypes = [GroupChatStartMessage]
;(GroupChatAgentActor.prototype._handleResponseMessage as any).targetTypes = [GroupChatResponseMessage]
;(GroupChatAgentActor.prototype._handleRequestMessage as any).targetTypes = [GroupChatRequestMessage]

// endregion GroupChatAgentActor

// region GroupChatManager

/**
 * A group chat manager that manages the flow of a group chat.
 */
@experimental
export abstract class GroupChatManager {
  currentRound: number = 0
  maxRounds?: number

  humanResponseFunction?: (chatHistory: ChatHistory) => ChatMessageContent | Promise<ChatMessageContent>

  /**
   * Check if the group chat should request user input.
   */
  abstract shouldRequestUserInput(chatHistory: ChatHistory): Promise<BooleanResult>

  /**
   * Check if the group chat should terminate.
   */
  async shouldTerminate(_chatHistory: ChatHistory): Promise<BooleanResult> {
    this.currentRound += 1

    if (this.maxRounds !== undefined) {
      return new BooleanResult(
        this.currentRound > this.maxRounds,
        this.currentRound > this.maxRounds ? 'Maximum rounds reached.' : 'Not reached maximum rounds.'
      )
    }
    return new BooleanResult(false, 'No maximum rounds set.')
  }

  /**
   * Select the next agent to speak.
   */
  abstract selectNextAgent(
    chatHistory: ChatHistory,
    participantDescriptions: Map<string, string>
  ): Promise<StringResult>

  /**
   * Filter the results of the group chat.
   */
  abstract filterResults(chatHistory: ChatHistory): Promise<MessageResult>
}

/**
 * A round-robin group chat manager.
 */
@experimental
export class RoundRobinGroupChatManager extends GroupChatManager {
  currentIndex: number = 0

  constructor(options?: {
    maxRounds?: number
    humanResponseFunction?: (chatHistory: ChatHistory) => ChatMessageContent | Promise<ChatMessageContent>
  }) {
    super()
    if (options?.maxRounds !== undefined) {
      this.maxRounds = options.maxRounds
    }
    if (options?.humanResponseFunction !== undefined) {
      this.humanResponseFunction = options.humanResponseFunction
    }
  }

  async shouldRequestUserInput(_chatHistory: ChatHistory): Promise<BooleanResult> {
    return new BooleanResult(false, 'The default round-robin group chat manager does not request user input.')
  }

  async selectNextAgent(
    _chatHistory: ChatHistory,
    participantDescriptions: Map<string, string>
  ): Promise<StringResult> {
    // Handle both Map and plain object
    const keys =
      participantDescriptions instanceof Map
        ? Array.from(participantDescriptions.keys())
        : Object.keys(participantDescriptions)
    const nextAgent = keys[this.currentIndex]
    this.currentIndex = (this.currentIndex + 1) % keys.length
    return new StringResult(nextAgent, 'Round-robin selection.')
  }

  async filterResults(chatHistory: ChatHistory): Promise<MessageResult> {
    return new MessageResult(
      chatHistory.messages[chatHistory.messages.length - 1],
      'The last message in the chat history is the result in the default round-robin group chat manager.'
    )
  }
}

// endregion GroupChatManager

// region GroupChatManagerActor

/**
 * A group chat manager actor.
 */
@experimental
export class GroupChatManagerActor extends ActorBase {
  private _manager: GroupChatManager
  private _internalTopicType: string
  private _chatHistory: ChatHistory
  private _participantDescriptions: Map<string, string>
  private _resultCallback?: (result: DefaultTypeAlias) => void | Promise<void>

  constructor(options: {
    manager: GroupChatManager
    internalTopicType: string
    participantDescriptions: Map<string, string>
    exceptionCallback: (exception: Error) => void
    resultCallback?: (result: DefaultTypeAlias) => void | Promise<void>
  }) {
    super('An actor for the group chat manager.', options.exceptionCallback)

    this._manager = options.manager
    this._internalTopicType = options.internalTopicType
    this._chatHistory = new ChatHistory()
    this._participantDescriptions = options.participantDescriptions
    this._resultCallback = options.resultCallback
  }

  @event()
  async _handleStartMessage(message: GroupChatStartMessage, ctx: MessageContext): Promise<void> {
    logger.debug(`${this.id}: Received group chat start message.`)
    if (message.body instanceof ChatMessageContent) {
      this._chatHistory.addMessage(message.body)
    } else if (Array.isArray(message.body) && message.body.every((m) => m instanceof ChatMessageContent)) {
      for (const m of message.body) {
        this._chatHistory.addMessage(m)
      }
    } else {
      throw new Error(`Invalid message body type: ${typeof message.body}. Expected DefaultTypeAlias.`)
    }

    await this._determineStateAndTakeAction(ctx.cancellationToken)
  }

  @event()
  async _handleResponseMessage(message: GroupChatResponseMessage, ctx: MessageContext): Promise<void> {
    if (message.body.role !== AuthorRole.USER) {
      this._chatHistory.addMessage(
        new ChatMessageContent({
          role: AuthorRole.USER,
          content: `Transferred to ${message.body.name}`,
        })
      )
    }
    this._chatHistory.addMessage(message.body)

    await this._determineStateAndTakeAction(ctx.cancellationToken)
  }

  private async _determineStateAndTakeAction(cancellationToken: CancellationToken): Promise<void> {
    try {
      // User input state
      const chatHistoryCopy = new ChatHistory([...this._chatHistory.messages])
      const shouldRequestUserInput = await this._manager.shouldRequestUserInput(chatHistoryCopy)
      if (shouldRequestUserInput.result && this._manager.humanResponseFunction) {
        logger.debug(`Group chat manager requested user input. Reason: ${shouldRequestUserInput.reason}`)
        const userInputMessage = await this._callHumanResponseFunction()
        this._chatHistory.addMessage(userInputMessage)
        await this.publishMessage(
          new GroupChatResponseMessage(userInputMessage),
          new TopicId(this._internalTopicType, this.id.key),
          { cancellationToken }
        )
        logger.debug('User input received and added to chat history.')
      }

      // Determine if the group chat should terminate
      const shouldTerminate = await this._manager.shouldTerminate(new ChatHistory([...this._chatHistory.messages]))
      if (shouldTerminate.result) {
        logger.debug(`Group chat manager decided to terminate the group chat. Reason: ${shouldTerminate.reason}`)
        if (this._resultCallback) {
          const result = await this._manager.filterResults(new ChatHistory([...this._chatHistory.messages]))
          result.result.metadata = result.result.metadata || {}
          result.result.metadata['termination_reason'] = shouldTerminate.reason
          result.result.metadata['filter_result_reason'] = result.reason
          await this._resultCallback(result.result)
        }
        return
      }

      // Select the next agent to speak if the group chat is not terminating
      const nextAgent = await this._manager.selectNextAgent(
        new ChatHistory([...this._chatHistory.messages]),
        this._participantDescriptions
      )
      logger.debug(
        `Group chat manager selected agent: ${nextAgent.result} on round ${this._manager.currentRound}. ` +
          `Reason: ${nextAgent.reason}`
      )

      await this.publishMessage(
        new GroupChatRequestMessage(nextAgent.result),
        new TopicId(this._internalTopicType, this.id.key),
        { cancellationToken }
      )
    } catch (error) {
      this._exceptionCallback(error as Error)
      logger.error(`Exception occurred in group chat manager ${this.id}:`, error)
      throw error
    }
  }

  private async _callHumanResponseFunction(): Promise<ChatMessageContent> {
    if (!this._manager.humanResponseFunction) {
      throw new Error('Human response function is not set.')
    }
    const result = this._manager.humanResponseFunction(new ChatHistory([...this._chatHistory.messages]))
    if (result instanceof Promise) {
      return await result
    }
    return result
  }
}

// Add message type registrations for the handlers
;(GroupChatManagerActor.prototype._handleStartMessage as any).targetTypes = [GroupChatStartMessage]
;(GroupChatManagerActor.prototype._handleResponseMessage as any).targetTypes = [GroupChatResponseMessage]

// endregion GroupChatManagerActor

// region GroupChatOrchestration

/**
 * A group chat multi-agent pattern orchestration.
 */
@experimental
export class GroupChatOrchestration<TIn, TOut> extends OrchestrationBase<TIn, TOut> {
  private _manager: GroupChatManager

  constructor(options: {
    members: Agent[]
    manager: GroupChatManager
    name?: string
    description?: string
    inputTransform?: (input: TIn) => DefaultTypeAlias | Promise<DefaultTypeAlias>
    outputTransform?: (output: DefaultTypeAlias) => TOut | Promise<TOut>
    agentResponseCallback?: (response: DefaultTypeAlias) => void | Promise<void>
    streamingAgentResponseCallback?: (response: StreamingChatMessageContent, isFinal: boolean) => void | Promise<void>
  }) {
    // Validate all members have descriptions
    for (const member of options.members) {
      if (!member.description) {
        throw new Error('All members must have a description.')
      }
    }

    super({
      members: options.members,
      name: options.name,
      description: options.description,
      inputTransform: options.inputTransform,
      outputTransform: options.outputTransform,
      agentResponseCallback: options.agentResponseCallback,
      streamingAgentResponseCallback: options.streamingAgentResponseCallback,
    })

    this._manager = options.manager
  }

  protected async _start(
    task: DefaultTypeAlias,
    runtime: CoreRuntime,
    internalTopicType: string,
    cancellationToken: CancellationToken
  ): Promise<void> {
    // Send start messages to all agent actors first
    const sendStartMessagePromises = this._members.map(async (agent) => {
      const targetActorId = await runtime.get(this._getAgentActorType(agent, internalTopicType))
      await runtime.sendMessage(new GroupChatStartMessage(task), targetActorId, { cancellationToken })
    })

    await Promise.all(sendStartMessagePromises)

    // Send the start message to the manager actor
    const targetActorId = await runtime.get(this._getManagerActorType(internalTopicType))
    await runtime.sendMessage(new GroupChatStartMessage(task), targetActorId, { cancellationToken })
  }

  protected async _prepare(
    runtime: CoreRuntime,
    options: {
      internalTopicType: string
      exceptionCallback: (exception: Error) => void
      resultCallback: (result: DefaultTypeAlias) => Promise<void>
    }
  ): Promise<void> {
    await this._registerMembers(runtime, options.internalTopicType, options.exceptionCallback)
    await this._registerManager(runtime, options.internalTopicType, options.exceptionCallback, options.resultCallback)
    await this._addSubscriptions(runtime, options.internalTopicType)
  }

  private async _registerMembers(
    runtime: CoreRuntime,
    internalTopicType: string,
    exceptionCallback: (exception: Error) => void
  ): Promise<void> {
    const registrationPromises = this._members.map((agent) =>
      GroupChatAgentActor.register(
        runtime as any,
        this._getAgentActorType(agent, internalTopicType),
        () =>
          new GroupChatAgentActor({
            agent,
            internalTopicType,
            exceptionCallback,
            agentResponseCallback: this._agentResponseCallback,
            streamingAgentResponseCallback: this._streamingAgentResponseCallback,
          })
      )
    )

    await Promise.all(registrationPromises)
  }

  private async _registerManager(
    runtime: CoreRuntime,
    internalTopicType: string,
    exceptionCallback: (exception: Error) => void,
    resultCallback?: (result: DefaultTypeAlias) => Promise<void>
  ): Promise<void> {
    await GroupChatManagerActor.register(runtime as any, this._getManagerActorType(internalTopicType), () => {
      const participantDescriptions = new Map<string, string>()
      for (const agent of this._members) {
        participantDescriptions.set(agent.name, agent.description!)
      }

      return new GroupChatManagerActor({
        manager: this._manager,
        internalTopicType,
        participantDescriptions,
        exceptionCallback,
        resultCallback,
      })
    })
  }

  private async _addSubscriptions(runtime: CoreRuntime, internalTopicType: string): Promise<void> {
    const subscriptions: TypeSubscription[] = []

    for (const agent of this._members) {
      subscriptions.push(new TypeSubscription(internalTopicType, this._getAgentActorType(agent, internalTopicType)))
    }
    subscriptions.push(new TypeSubscription(internalTopicType, this._getManagerActorType(internalTopicType)))

    await Promise.all(subscriptions.map((sub) => runtime.addSubscription(sub)))
  }

  private _getAgentActorType(agent: Agent, internalTopicType: string): string {
    return `${agent.name}_${internalTopicType}`
  }

  private _getManagerActorType(internalTopicType: string): string {
    return `GroupChatManagerActor_${internalTopicType}`
  }
}

// endregion GroupChatOrchestration
